### PR TITLE
login facebook

### PR DESCRIPTION
**Pull request by tri-et**

### CODE REVIEW

File: server/schema/customer/def.js [hunk 1]
```diff
@@ -24,14 +24,20 @@ const Customer = `
...
     username: String!
     password: String!
     passwordConfirm: String!
-    fullName: String,
+    name: String,
     phone: String
     address: String
+    type: String
+  }
+  type RegFbRes {
```
Comment (hemvanh) [inline server/schema/customer/def.js]: RegFbRes  -> RegisterFbResponse

File: server/schema/customer/def.js [hunk 2]
```diff
@@ -24,14 +24,20 @@ const Customer = `
   input LoginInput {
     username: String!
     password: String!
+    type: String
   }
   input RegInput {
```
Comment (hemvanh) [inline server/schema/customer/def.js]: RegInput -> RegisterInput

File: server/schema/customer/res.js [hunk 3]
```diff
@@ -37,7 +37,7 @@ const resolvers = {
       )
     },
     async reg(_, {input}) {
```
Comment (hemvanh) [inline server/schema/customer/res.js]: reg -> register

File: server/schema/customer/res.js [hunk 4]
```diff
@@ -59,6 +59,54 @@ const resolvers = {
         )
       })
     },
+    async loginFB(_, {input}) {
```
Comment (hemvanh) [inline server/schema/customer/res.js]: loginFB -> loginFb

File: server/schema/customer/res.js [hunk 5]
```diff
@@ -59,6 +59,54 @@ const resolvers = {
...
+          add: user.add,
+          phone: user.phone,
+          balance: user.balance,
+          points: user.points,
+        },
+        process.env.JWT_SECRET,
+        {expiresIn: '1y'}
+      )
+    },
+    async regFB(_, {input}) {
```
Comment (hemvanh) [inline server/schema/customer/res.js]: regFB -> registerFb

File: server/schema/customer/res.js [hunk 6]
```diff
@@ -59,6 +59,54 @@ const resolvers = {
...
+      var user = await Customer.findOne({where: {username: input.username, type: input.type}})
+      var msgRes = ''
+      if (user) {
+        msgRes = 'Account is existed!'
+        input.id = user.id
+      } else {
+        msgRes = 'Regitered Successfully!'
+      }
+      input.password = bcrypt.hashSync(input.password, SALT)
```
Comment (hemvanh) [inline server/schema/customer/res.js]: why need password ?

File: server/schema/customer/res.js [hunk 7]
```diff
@@ -59,6 +59,54 @@ const resolvers = {
...
+      return await Customer.upsert(input).then(async function() {
+        user = await Customer.findOne({where: {username: input.username}})
```
Comment (hemvanh) [inline server/schema/customer/res.js]: user already available above, why need to findOne again?

File: server/schema/index.js [hunk 8]
```diff
@@ -29,6 +29,8 @@ const RootMutation = `
   type RootMutation {
     login(input:LoginInput): String
     reg(input:RegInput): String
+    loginFB(input:LoginInput): String
+    regFB(input:RegInput): RegFbRes
```
Comment (hemvanh) [inline server/schema/index.js]: regFB(input:RegInput): RegFbRes
-> registerFb(input:RegisterInput): RegisterFbResponse
*note: aslways use "Fb"  dont mix with "FB"

File: src/index.template.html [hunk 9]
```diff
@@ -13,6 +13,7 @@
   <link rel="icon" href="statics/quasar-logo.png" type="image/x-icon">
   <link rel="icon" type="image/png" sizes="32x32" href="statics/icons/favicon-32x32.png">
   <link rel="icon" type="image/png" sizes="16x16" href="statics/icons/favicon-16x16.png">
+  <script src="statics/fb.js"></script>
```
Comment (hemvanh) [inline src/index.template.html]: fb.js -> fb-api-v1.2.3.js

File: src/pages/Login.vue [hunk 10]
```diff
@@ -67,6 +68,36 @@ export default {
         }
       )
     },
+    loginFB() {
```
Comment (hemvanh) [inline src/pages/Login.vue]: 1)why need to seperate into 2 functions: loginCustumerFB() , loginFB()
can put "window.FB.login(res => { ... into vuex action loginFB ?
2)same as above for Register
3)wrong Custumer -> Customer , FB -> Fb

File: src/store/customer/actions.js [hunk 11]
```diff
@@ -51,3 +51,52 @@ export function regCustomer({commit}, payload) {
       commit('setIsLoading', false)
     })
 }
+
+export function loginCustumerFB({commit}, payload) {
```
Comment (hemvanh) [inline src/store/customer/actions.js]: why use loginCustumerFB, can just use loginFb(), which call to the mutation loginFb

File: src/store/customer/actions.js [hunk 12]
```diff
@@ -51,3 +51,52 @@ export function regCustomer({commit}, payload) {
...
+        _alert(`Logged In Successfully!`, 'positive')
+        this.$router.push('/customer')
+      }
+    })
+    .catch(err => {
+      _procError(err)
+    })
+}
+
+export function regCustomerFB({commit}, payload) {
```
Comment (hemvanh) [inline src/store/customer/actions.js]: regCustomerFB -> registerFb

File: src/store/customer/actions.js [hunk 13]
```diff
@@ -51,3 +51,52 @@ export function regCustomer({commit}, payload) {
...
+  _post(
+    payload,
+    `mutation ($input: RegInput) {
+      regFB(input: $input) {
```
Comment (hemvanh) [inline src/store/customer/actions.js]: regFB -> registerFb

**Review by hemvanh — CHANGES_REQUESTED** [review]
please fix all the issues as commented

File: server/schema/customer/res.js (hunk 7 above)
Comment (tri-et) [reply server/schema/customer/res.js]: because we need user's id just registered for set token

File: server/schema/customer/res.js [hunk 14]
```diff
@@ -59,6 +59,54 @@ const resolvers = {
...
+      var user = await Customer.findOne({where: {username: input.username, type: input.type}})
+      var msgRes = ''
+      if (user) {
+        msgRes = 'Account is existed!'
+        input.id = user.id
+      } else {
+        msgRes = 'Regitered Successfully!'
+      }
+      input.password = bcrypt.hashSync(input.password, SALT)
+      return await Customer.upsert(input).then(async function() {
```
Comment (hemvanh) [inline server/schema/customer/res.js:71]: upsert -> insert to be safe, coz update should not be happening here

File: server/schema/customer/res.js [hunk 15]
```diff
@@ -59,6 +59,54 @@ const resolvers = {
         )
       })
     },
+    async loginFB(_, {input}) {
+      const user = await Customer.findOne({where: {username: input.username}})
```
Comment (hemvanh) [inline server/schema/customer/res.js:56]: refactor this part into a reusable async function
ex:      async generateLoginJwt(username)  -> return jwt.sign(....)

**Review by hemvanh — CHANGES_REQUESTED** [review]
refactor jwt part to generateLoginJwt()
be careful when use upsert

File: server/schema/customer/mutations.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
+import jwt from 'jsonwebtoken'
```
Comment (hemvanh) [inline server/schema/customer/mutations.js]: why need to put in mutations.js ?

File: server/schema/customer/res.js [hunk 17]
```diff
@@ -23,18 +23,7 @@ const resolvers = {
...
-          name: user.name,
-          add: user.add,
-          phone: user.phone,
-          balance: user.balance,
-          points: user.points,
-        },
-        process.env.JWT_SECRET,
-        {expiresIn: '1y'}
-      )
+      return generateLoginJwt(Customer, input)
```
Comment (hemvanh) [inline server/schema/customer/res.js]: un-necessary passing of Customer

**Review by hemvanh — CHANGES_REQUESTED** [review]
refactor 1 more time

File: server/schema/customer/res.js [hunk 18]
```diff
@@ -1,9 +1,26 @@
 import {Customer} from '../../models'
 import GraphQLDate from 'graphql-date'
 import bcrypt from 'bcrypt'
+import jwt from 'jsonwebtoken'
 import {_auth} from '../../util'
-import {generateLoginJwt} from './mutations'
 const SALT = 10
+async function generateLoginJwt(input, msg = '') {
```
Comment (hemvanh) [inline server/schema/customer/res.js]: input -> customer , best send in "username" only
** and why do we need to pass-in object, while we only need username for searching ?
for the naming scheme "input" is just too unambiguous!

**Review by hemvanh — CHANGES_REQUESTED** [review]
simplify the params passing for generateLoginJwt()

**Review by hemvanh — APPROVED** [review]

**Pull request merged**